### ISSUE TITLE
Function type-checking and casting doesn't respect JS interop class inheritance in DDC

### ISSUE THREAD

**Issue by greglittlefield-wf** [dart-lang/sdk#43939]
When upcasting a function type by loosening the type of an argument, whose type also happens to be a JS interop type, the type-check or cast fails.

This only happens in the DDC, and only affects certain use-cases, since the DDC doesn't emit type-checks everywhere.

I've tried and reproduced on Dart 2.11.0-240.0.dev, 2.9.3, and 2.7.2.

Here's a reduced test case, using an implicit cast to force a type-check. Changing it to an explicit cast also results in the same issue.
```dart
@JS()
library test;

import 'package:js/js.dart';

@JS()
@anonymous
class A {}

@JS()
@anonymous
class B extends A {}

main() {
  dynamic function = (A event) {};
  // throws: Expected a value of type '(B) => dynamic', but got one of type '(A) => Null'
  dynamic Function(B) function2 = function; 
}
```

Full error:
```
require.js:143 Uncaught Error: Expected a value of type '(B) => dynamic', but got one of type '(A) => Null'
    at Object.throw_ [as throw] (dart_sdk.js:3880)
    at Object.castError (dart_sdk.js:3840)
    at dart.FunctionType.new.as (dart_sdk.js:5498)
    at dart.FunctionType.new._check (dart_sdk.js:5501)
    at Object.main$ [as main] (main.ddc.js:13)
```

I hit this issue when using `@anonymous` classes, but I also get a similar error when using non-`@anonymous` classes:

```js
// JS
function A() {}
function B() {}
B.prototype = new A();
```
```dart
// Dart
@JS()
library test;

import 'package:js/js.dart';

@JS()
class A {}

@JS()
class B extends A {}

main() {
  dynamic function = (A event) {};
  // throws: Uncaught Error: Expected a value of type '(JSObject<B>) => dynamic', but got one of type '(JSObject<A>) => Null'
  dynamic Function(B) function2 = function;
}
```

The same code doesn't exhibit the issue if you're using only pure Dart classes, so I'd expect it to work for JS classes:
```dart
class A {}
class B extends A {}

main() {
  dynamic function = (A event) {};
  dynamic Function(B) function2 = function; // no error
}
```

**Comment by greglittlefield-wf** [comment]
It looks like a potential workaround to this issue is to use non-anonymous JS classes that don't have native types:

```dart
@JS('someNameThatDoesNotExist_A')
class A {}

@JS('someNameThatDoesNotExist_B')
class B extends A {}
```

The type check-passes, presumably because type checking is skipped as a result of the type not being found
```
dart_sdk.js:4850 Cannot find native JavaScript type (someNameThatDoesNotExist_B) for type check
  _warn			@ dart_sdk.js:4850
  [_getRawJSType]	@ dart_sdk.js:5771
  rawJSTypeForCheck	@ dart_sdk.js:5780
  _isInterfaceSubtype	@ dart_sdk.js:5313
  _isSubtype		@ dart_sdk.js:5280
  _isFunctionSubtype	@ dart_sdk.js:5118
  _isSubtype		@ dart_sdk.js:5310
  isSubtypeOf		@ dart_sdk.js:5186
  is			@ dart_sdk.js:6102
  as			@ dart_sdk.js:5897
  main$			@ main.ddc.js:14
dart_sdk.js:4850 Cannot find native JavaScript type (someNameThatDoesNotExist_A) for type check
  ...
```

...but that seems pretty hacky.

**Comment by sigmundch** [comment]
Thanks for the report. This is indeed a very tricky area of js-interop. For now js-interop types in compounded types (generics and function types) are likely behaving inconsistently between DDC and dart2js. It so happens that @srujzs is currently looking into this same set of issues, but it will likely take us some time to iron out the details.

**Comment by greglittlefield-wf** [comment]
Awesome to hear you guys are already looking into it; thanks for the info!

Do you have any thoughts on the viability of that workaround?

**Comment by greglittlefield-wf** [comment]
It appears this issue was fixed in Dart 2.14.0. In that version and later, I can no longer reproduce!

Seems likely related to this changelog entry:
> **Dart Dev Compiler (DDC)**
> - **Breaking Change #44154**: Subtyping relations of package:js classes have been changed to be more correct and consistent with Dart2JS. Like anonymous classes, non-anonymous classes will no longer check the underlying type in DDC. The internal type representation of these objects have changed as well, which will affect the toString value of these types.

Closing this issue, thanks!
